server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -222,46 +229,87 @@ protected void runTasks(TaskInputs taskInputs) {
                 }
 
                 logger.debug("publishing cluster state version [{}]", newClusterState.version());
-                try {
-                    clusterStatePublisher.accept(clusterChangedEvent, taskOutputs.createAckListener(threadPool, newClusterState));
-                } catch (Discovery.FailedToCommitClusterStateException t) {
-                    final long version = newClusterState.version();
-                    logger.warn(() -> new ParameterizedMessage(
-                            "failing [{}]: failed to commit cluster state version [{}]", summary, version), t);
-                    taskOutputs.publishingFailed(t);
-                    return;
-                }
+                publish(clusterChangedEvent, taskOutputs, startTimeNS);
+            } catch (Exception e) {
+                handleException(summary, startTimeNS, newClusterState, e);
+            }
+        }
+    }
+
+    protected void publish(ClusterChangedEvent clusterChangedEvent, TaskOutputs taskOutputs, long startTimeNS) throws Exception {
+        CompletableFuture<Void> fut = new CompletableFuture<>();
+        clusterStatePublisher.publish(clusterChangedEvent, new ActionListener<Void>() {
```

Comment:
The reason I did not choose `PlainActionFuture` was because it asserts that we're not blocking on the MasterServiceUpdateThread (which this future deliberately does). Unfortunately, `CompletableFuture` has other problems (https://github.com/elastic/elasticsearch/pull/32512#issuecomment-409995328), so I've gone back to `PlainActionFuture`, but added a hook that allows to disable checking some of the assertions, see 526511d546f